src/plugins/network.ts

```diff
@@ -23,336 +23,217 @@
+    const request = response.request();
+    const networkEntry = this._findNetworkEntry(request);
+    if (!networkEntry) return;
+
+    const server = await response.serverAddr();
+    const responseHeaders = await response.allHeaders();
+    const mimeType = responseHeaders['content-type']
+      ? responseHeaders['content-type'].split(';')[0]
+      : 'unknown';
+
+    const requestHeaders = await request.allHeaders();
+
+    networkEntry.request.headers = requestHeaders;
+    networkEntry.request.referrer = requestHeaders?.referer;
+    networkEntry.response = {
+      url: response.url(),
+      statusCode: response.status(),
+      statusText: response.statusText(),
+      headers: responseHeaders,
+      mimeType,
+      redirectURL: networkEntry.response.redirectURL,
+      securityDetails: await response.securityDetails(),
+      remoteIPAddress: server?.ipAddress,
+      remotePort: server?.port,
+    };
+    networkEntry.status = response.status();
+    networkEntry.responseReceivedTime = epochTimeInMills();
+  }
+
+  private async _onRequestCompleted(request: Request) {
+    const networkEntry = this._findNetworkEntry(request);
+    if (!networkEntry) return;
+
+    networkEntry.loadEndTime = epochTimeInMills();
+
+    // For aborted/failed requests sizes does not exist
+    const sizes = await request.sizes().catch(() => {});
+    if (sizes) {
+      networkEntry.request.bytes =
+        sizes.requestHeadersSize + sizes.requestBodySize;
+      networkEntry.request.body = {
+        bytes: sizes.requestBodySize,
+      };
+      networkEntry.response.bytes =
+        sizes.responseHeadersSize + sizes.responseBodySize;
+      networkEntry.response.body = {
+        bytes: sizes.responseBodySize,
+      };
+      networkEntry.transferSize = sizes.responseBodySize;
     }
-    Object.assign(record, {
-      status: response.status,
-      response,
-      responseReceivedTime: timestamp,
-      transferSize: response.encodedDataLength,
-    });
-  }
 
-  _onLoadingFinished(event: Protocol.Network.loadingFinishedPayload) {
-    const { requestId, timestamp, encodedDataLength } = event;
-    this._requestCompleted(requestId, timestamp, encodedDataLength);
-  }
+    const timing = request.timing();
+    const { loadEndTime, requestSentTime } = networkEntry;
+    networkEntry.timings = {
+      blocked: -1,
+      dns: -1,
+      ssl: -1,
+      connect: -1,
+      send: -1,
+      wait: -1,
+      receive: -1,
+      total: -1,
+    };
 
-  _onLoadingFailed(event: Protocol.Network.loadingFailedPayload) {
-    const { requestId, timestamp } = event;
-    this._requestCompleted(requestId, timestamp);
-  }
+    const firstPositive = (numbers: number[]) => {
+      for (let i = 0; i < numbers.length; ++i) {
+        if (numbers[i] > 0) {
+          return numbers[i];
+        }
+      }
+      return null;
+    };
+    const roundMilliSecs = (value: number): number => {
+      return ((value * 1000) | 0) / 1000;
```

Comment:
Is there a more readable alternative to use bitwise operators here?

Reply:
Math.floor, I can switch to that.